source/grid/manifold_lib.cc

```diff
@@ -1209,35 +1249,61 @@ template <int dim, int spacedim>
-  // this class does not throw exception upon failure as those are relatively
-  // expensive and failure occurs quite regularly in the implementation of the
-  // compute_chart_points method.
+  // run quasi-Newton iteration with a combination of finite differences for
+  // the exact Jacobian and "Broyden's good method". As opposed to the various
+  // mapping implementations, this class does not throw exception upon failure
+  // as those are relatively expensive and failure occurs quite regularly in
+  // the implementation of the compute_chart_points method.
   Tensor<1,spacedim> residual = point - compute_transfinite_interpolation(*cell, chart_point,
                                 coarse_cell_is_flat[cell->index()]);
-  const double tolerance = 1e-21 * cell->diameter() * cell->diameter();
+  const double tolerance = 1e-21 * Utilities::fixed_power<2>(cell->diameter());
   double residual_norm_square = residual.norm_square();
+  DerivativeForm<1,dim,spacedim> inv_grad;
   for (unsigned int i=0; i<100; ++i)
     {
       if (residual_norm_square < tolerance)
-        return chart_point;
+        {
+          // do a final update of the point with the last available Jacobian
+          // information. The residual is close to zero due to the check
+          // above, but me might improve some of the last digits by a final
+          // Newton-like step with step length 1
+          Tensor<1,dim> update;
+          for (unsigned int d=0; d<spacedim; ++d)
+            for (unsigned int e=0; e<dim; ++e)
+              update[e] += inv_grad[d][e] * residual[d];
+          return chart_point + update;
+        }
 
-      // if the determinant is zero, the mapping is not invertible and we are
-      // outside the valid chart region
-      DerivativeForm<1,dim,spacedim> grad = push_forward_gradient(cell, chart_point);
-      if (grad.determinant() <= 0.0)
-        return outside;
-      DerivativeForm<1,dim,spacedim> inv_grad = grad.covariant_form();
+      // every 8 iterations, including the first time around, we create an
+      // approximation of the Jacobian with finite differences. Broyden's
+      // method usually does not need more than 5-8 iterations, but sometimes
+      // we might have had a bad initial guess and then we can accelerate
+      // convergence considerably with getting the actual Jacobian rather than
+      // using secant-like methods (one gradient calculation in 3D costs as
+      // much as 3 more iterations). this usually happens close to convergence
+      // and one more step with the finite-differenced Jacobian leads to
+      // convergence
+      if (i%8 == 0)
+        {
+          // if the determinant is zero, the mapping is not invertible and we are
+          // outside the valid chart region
+          DerivativeForm<1,dim,spacedim> grad
+            = push_forward_gradient(cell, chart_point,
+                                    Point<spacedim>(point-residual));
+          if (grad.determinant() <= 0.0)
```

Comment:
Dumb question: why can the determinant of the push forward derivative not be negative?

Reply:
I will add a comment.